tests/001-handshake.cpp

```diff
@@ -283,4 +283,122 @@ namespace oxen::quic::test
         CHECK(server_established.wait());
         CHECK(client_ci->is_validated());
     };
+
+    TEST_CASE("001 - Handshaking: Defer", "[001][defer][quietclose]")
+    {
+        auto client_established = callback_waiter{[](connection_interface&) {}};
+        auto server_established = callback_waiter{[](connection_interface&) {}};
+
+        Network test_net{};
+
+        std::shared_ptr<connection_interface> server_ci, client_ci;
+
+        // Instead of using randomly generated seeds and pubkeys, hardcoded strings are used to deterministically
+        // produce the same test result. The key verify callback compares the pubkeys in lexicographical order,
+        // deferring to the connetion initiated by the pubkey that appears first in said order.
+        const std::string C_SEED = "468e7ed2cd914ca44568e7189245c7b8e5488404fc88a4019c73b51d9dbc48a5"_hex;
+        const std::string C_PUBKEY = "626136fe40c8860ee5bdc57fd9f15a03ef6777bb9237c18fc4d7ef2aacfe4f88"_hex;
+        const std::string S_SEED = "fefbb50cdd4cde3be0ae75042c44ff42b026def4fd6be4fb1dc6e81ea0480c9b"_hex;
+        const std::string S_PUBKEY = "d580d5c68937095ea997f6a88f07a86cdd26dfa0d7d268e80ea9bbb5f3ca0304"_hex;
+
+        auto client_tls = GNUTLSCreds::make_from_ed_keys(C_SEED, C_PUBKEY);
+        auto server_tls = GNUTLSCreds::make_from_ed_keys(S_SEED, S_PUBKEY);
+
+        auto defer_hook = [](const std::string& incoming,
+                             const std::string& local,
+                             const std::string& remote,
+                             std::shared_ptr<connection_interface> local_outbound) -> bool {
+            REQUIRE(incoming == remote);
+
+            auto rv = std::strcmp(incoming.c_str(), local.c_str());
+
+            // The pubkeys definitely should not be the same
+            REQUIRE(rv != 0);
+
+            // If the LHS parameter to std::strcmp appears FIRST in lexicographical order, then rv < 0. As a result,
+            // if the incoming pubkey appears BEFORE the server pubkey in lexicographical order, we will defer to the
+            // connection initiated by the remote -- ergo we will ACCEPT this connection and mark the local endpoint's
+            // connection as "die silently" (close w/o executing any callbacks or writing any close packets). Else, we
+            // will REJECT the incoming connection and defer to the local endpoint's outgoing connection
+            auto defer_to_incoming = rv < 0;
```

Comment:
```suggestion
            // The pubkeys definitely should not be the same
            REQUIRE(incoming != local);

            // If the LHS parameter to std::strcmp appears FIRST in lexicographical order, then rv < 0. As a result,
            // if the incoming pubkey appears BEFORE the server pubkey in lexicographical order, we will defer to the
            // connection initiated by the remote -- ergo we will ACCEPT this connection and mark the local endpoint's
            // connection as "die silently" (close w/o executing any callbacks or writing any close packets). Else, we
            // will REJECT the incoming connection and defer to the local endpoint's outgoing connection
            auto defer_to_incoming = incoming < local;
```
No need to go to C functions for lexicographic comparison.

Reply:
(The string can also contain `\x00` characters, which the strcmp won't deal with correctly but std::string `<` will)